# PPTEs

**PPTEs** (Pangenome Polymorphic Transposable Elements simulation toolkit) is a Python toolkit for simulating polymorphic transposable elements (pTEs) in genomes. It supports generating TE sequence pools, simulating real pTEs, generating simulated VCF and genome sequences, and comparing predicted VCFs from other TE genotyping tools with the simulated VCF file.

---

## Installation

You can install PPTEs from source using `pip`:

```bash
pip install PPTE
```
##  Dependencies
The package depends on the following Python libraries:
- Python 3.8+
- numpy
- biopython
- pysam

## Quick start
**1. Simulate 100 pTE from known TE insertions and deletions**
```bash
ppte TEreal --knownINS MEI_Callset_GRCh38.ALL.20241211.fasta \
--knownDEL rmsk_chr21.txt \
--CHR 21 --nTE 100
```
- `MEI_Callset_GRCh38.ALL.20241211.fasta` is known pTE insertion, from paper [Logsdon, G.A. et al. Nature, 2025](https://www.nature.com/articles/s41586-025-09140-6)  
- `rmsk_chr21.txt` is the known repeats in Hg38-Chr21, downloaded from UCSC hgTables.  

**2. Simulate 100 genomes with 100 pTE**  
```bash
ppte simulate --ref chr21.fa \
--bed real.bed --num 100 \
--pool MEI_Callset_GRCh38.ALL.20241211.fasta
```
- `chr21.fa` is the sequence of Hg38-Chr21  
- `MEI_Callset_GRCh38.ALL.20241211.fasta` contains known pTE insertion sequences  
- `real.bed` is the position of pTE events that generated from `ppte TEreal`  

## Flowchart
![flowchart](https://github.com/JanMiao/PPTE/blob/main/flowchart.png)
- The TE deletion information can be obtained from [UCSC annotaion file (.txt)](https://genome.ucsc.edu/cgi-bin/hgTables) or [repeatmasker annotation (.out)  ](https://www.repeatmasker.org/genomicDatasets/RMGenomicDatasets.html)
- The TE insertion position can be obtained from inbuilt dataset (MEI_Callset_GRCh38.ALL.20241211.fasta). Any TE insertion sequence is acceptable , as long as the sequence ID follows the naming format **CHR-POS-ID**, e.g., **chr1-683234-AluSp**

## Usage
PPTEs provides four main command-line subcommands:
```bash
ppte <subcommand> [options]
```
### 1. TEpool

Generate a TE sequence pool from TE consensus sequences.

**Required arguments:**
- `consensus` : Path to TE consensus FASTA file

**Optional arguments:**
- `num` : Number of TE sequences to generate (default: 1000)
- `outprefix` : Output prefix for TE pool FASTA (default: TEpool)
- `snp-rate` : SNP mutation rate per base (default: 0.02)
- `indel-rate` : INDEL mutation rate per base (default: 0.005)
- `ins-ratio` : Proportion of indels that are insertions (default: 0.4)
- `indel-geom-p` : Geometric distribution parameter for indel lengths (default: 0.7)
- `truncated-ratio` : Proportion of sequences to truncate (default: 0.3)
- `truncated-max-length` : Maximum proportion of sequence to truncate (default: 0.5)
- `polyA-ratio` : Proportion of sequences to add polyA tail (default: 0.8)
- `polyA-min` : Minimum polyA length (default: 5)
- `polyA-max` : Maximum polyA length (default: 20)
- `seed` : Random seed (default: None)
- `verbose` : Disable verbose logging (default: True)

**Example**:
```bash
# Obtain 1000 TE by introducing random SNPs and INDELs,
# as well as 3' polyA tail and 5' truncation,
# into the TE consensus sequences
ppte TEpool --consensus consensus.fa --num 1000
```

### 2. TEreal
Automatically generate pTE positions from RepeatMasker or UCSC repeat annotations.

**Required arguments:**  
-`knownINS` : Known TE insertion file (FASTA)  
-`knownDEL` : Known TE deletion file (from RepeatMasker `.out` or UCSC `.txt`)  
-`CHR` : Chromosome used to simulate pTE  

**Optional arguments:**  
-`outprefix` : Output prefix for BED file (default: real)  
-`nTE` : Number of pTE insertions (default: 500)  
-`ins-ratio` : Proportion of insertion events (default: 0.4)  
-`seed` : Random seed (default: None)  
-`verbose` : Disable verbose logging (default: True)  

**Example**:
```bash
# simulate 500 pTE events from known TE insertion and deletion
ppte TEreal --knownINS MEI.fa --knownDEL rmsk_chr21.txt --CHR chr21 --nTE 500
```

### 3. simulate
Simulate pTE insertions/deletions and generate VCF and modified genome FASTA.

**Required arguments:**
-`ref` : Reference genome FASTA  
-`te-pool` : TE pool FASTA  
-`bed` : BED file of TE positions (can be generated by `ppte TEreal`)  
-`num` : Number of simulated genomes  

**Optional arguments:**  
-`outprefix` : Output prefix (default: Sim)  
-`af-min / --af-max` : Min/max allele frequency (default: 0.1/0.9)  
-`tsd-min / --tsd-max` : Min/max TSD length (default: 5/20)  
-`sense-strand-ratio` : Proportion of sense-strand insertions (default: 0.5)  
-`seed` : Random seed (default: None)  
-`verbose` : Disable verbose logging (default: True)  

**Example**:
```bash
# simulate 2 genome with known pTEs in real.bed
ppte simulate --ref chr21.fa --te-pool TEpool_out.fasta --bed real.bed --num-genomes 2
```

### 4. compare
Compare predicted VCF to the simulated VCF.

**Required arguments:**  
-`truth` : Ground truth VCF (generated by `ppte simulate`)  
-`pred` : Predicted VCF  
-`outprefix` : the Matched variants in two VCF file  
-`truthID` : Sample ID in truth VCF  
-`predID` : Sample ID in predicted VCF  

**Optional arguments:**  
-`nHap` : Ploidy level, e.g., 2 for Humans  (default: 2)  
-`max_dist` : Maximum allowed distance for variant matching (default: 100 bp)  

**Example**:
```bash
ppte compare --truth sim.vcf --pred variants.vcf --truthID Hap1_Hap2 --predID Sample
```
In simulation files, genomes are named Hap1, Hap2, etc.; for polyploids, combine haplotype IDs with `_` for one individual.  

